src/ui.rs

```diff
@@ -3047,40 +3047,60 @@ impl Component for SectionLineView<'_> {
     }
 
     fn draw(&self, viewport: &mut Viewport<Self::Id>, x: isize, y: isize) {
-        const NEWLINE_ICON: &str = "⏎";
-        let Self { line_key: _, inner } = self;
+        fn replacement_characters(character: char) -> Option<&'static str> {
+            match character {
+                // Characters end up writing over each-other and end up
+                // displaying incorrectly if ignored. Replacing tabs
+                // with a known length string fixes the issue for now.
+                '\t' => Some("→   "),
+                '\n' => Some("⏎"),
+                _ => None,
+            }
+        }
+
+        /// Split the line into a sequence of [`Span`]s where control characters are
+        /// replaced with styled [`Span`]'s and push them to the [`spans`] argument.
+        fn push_spans_from_line<'line>(line: &'line str, spans: &mut Vec<Span<'line>>) {
+            const CONTROL_CHARACTER_STYLE: Style = Style::new().fg(Color::DarkGray);
+
+            let mut last_index = 0;
+            // Find index of the start of each character to replace
+            for (idx, char) in line.match_indices(|char| replacement_characters(char).is_some()) {
+                // Push the string leading up to the character and the styled replacment string
+                if let Some(replacment_string) =
+                    char.chars().next().and_then(replacement_characters)
+                {
+                    spans.push(Span::raw(&line[last_index..idx]));
+                    spans.push(Span::styled(replacment_string, CONTROL_CHARACTER_STYLE));
```

Comment:
note: right now, this makes every control character (tab and newline) render with the background as grey when it's "selected", I _think_ that's fine? but I wanted to call it out to make sure.

<img width="674" alt="image" src="https://github.com/arxanas/scm-record/assets/6913198/4068635b-42ca-4597-9222-804fd9673126">
<img width="674" alt="image" src="https://github.com/arxanas/scm-record/assets/6913198/3b7bcf7e-9799-453e-908c-f05effe45b26">

Reply:
Oh I just noticed, should the ellipsis be rendered with a different style?

Reply:
Not sure what the best behavior is, but this is probably fine for now!